framework/src/utils/MooseUtils.C

```diff
@@ -1194,6 +1197,72 @@ prettyCppType(const std::string & cpp_type)
   return s;
 }
 
+template <typename Consumers>
+std::deque<MaterialBase *>
+buildRequiredMaterials(const Consumers & mat_consumers,
+                       const std::vector<std::shared_ptr<MaterialBase>> & mats,
+                       const bool allow_stateful)
+{
+  std::deque<MaterialBase *> required_mats;
+
+  std::unordered_set<unsigned int> needed_mat_props;
+  for (const auto & consumer : mat_consumers)
+  {
+    const auto & mp_deps = consumer->getMatPropDependencies();
+    needed_mat_props.insert(mp_deps.begin(), mp_deps.end());
+  }
+
+  // A predicate of calling this function is that these materials come in already sorted by
+  // dependency with the front of the container having no other material dependencies and following
+  // materials potentially depending on the ones in front of them. So we can start at the back and
+  // iterate forward checking whether the current material supplies anything that is needed, and if
+  // not we discard it
```

Comment:
Is this design feature taking of care of mortar properties provided by a material that uses data from another material? Would we error out if the second material happens to be stateful?

Reply:
Never mind the second question. Clearly spelled out below: https://github.com/idaholab/moose/blob/4a5d53f1ed5db3667a3358c55ba933359fada36d/framework/src/utils/MooseUtils.C#L1241.

Reply:
And the magic for the first question appears to take place here:
https://github.com/idaholab/moose/blob/4a5d53f1ed5db3667a3358c55ba933359fada36d/framework/src/utils/MooseUtils.C#L1246-L1248